File: memcache.java
public class Memcache {

    private Map<Integer, int[]> map;
    public Memcache() {
        // Initialize your data structure here.
        map = new HashMap<Integer, int[]>();
    }

    public int get(int curtTime, int key) {
        // Write your code here
        int[] temp;
        if (map.containsKey(key)) {
            temp = map.get(key);
        } else {
            return Integer.MAX_VALUE;
        }

        if (curtTime >= temp[1] && temp[1] != 0) {
            return Integer.MAX_VALUE;
        } else {
            return temp[0];
        }
    }

    public void set(int curtTime, int key, int value, int ttl) {
        // Write your code here
        int expire = curtTime + ttl;
        if (ttl == 0) {
            expire = 0;
        }
        int[] val = {value, expire};
        map.put(key, val);
    }

    public void delete(int curtTime, int key) {
        // Write your code here
        if (!map.containsKey(key)) {
            return;
        } else {
            map.remove(key);
        }
    }

    public int incr(int curtTime, int key, int delta) {
        // Write your code here
        if (!map.containsKey(key)) {
            return Integer.MAX_VALUE;
        }
        if (curtTime >= map.get(key)[1] && map.get(key)[1] != 0) {
            return Integer.MAX_VALUE;
        } else {
            int temp = map.get(key)[0];
            temp += delta;
            int[] val = {temp, map.get(key)[1]};
            map.put(key, val);
            return temp;
        }
    }

    public int decr(int curtTime, int key, int delta) {
        // Write your code here
        if (!map.containsKey(key)) {
            return Integer.MAX_VALUE;
        }
        if (curtTime >= map.get(key)[1] && map.get(key)[1] != 0) {
            return Integer.MAX_VALUE;
        } else {
            int temp = map.get(key)[0];
            temp -= delta;
            int[] val = {temp, map.get(key)[1]};
            map.put(key, val);
            return temp;
        }
    }
}

File: memcache.py
class Resource:
    def __init__(self, value, expired):
        self.value = value
        self.expired = expired

INT_MAX = 0x7fffffff

class Memcache:

    def __init__(self):
        # initialize your data structure here.
        self.client = {}
    
    # @param {int} curtTime an integer
    # @param {int} key an integer
    # @return an integer
    def get(self, curtTime, key):
        # Write your code here
        if key not in self.client:
            return INT_MAX
        res = self.client.get(key)
        if res.expired >= curtTime or res.expired == -1:
            return res.value
        return INT_MAX

    # @param {int} curtTime an integer
    # @param {int} key an integer
    # @param {int} integer
    # @param {int} ttl an integer
    # @return nothing
    def set(self, curtTime, key, value, ttl):
        # Write your code here
        if ttl:
            res = Resource(value, curtTime + ttl - 1)
        else:
            res = Resource(value, -1)
        self.client[key] = res
        
    # @param {int} curtTime an integer
    # @param {int} key an integer
    # @return nothing
    def delete(self, curtTime, key):
        # Write your code here
        if key not in self.client:
            return
        del self.client[key]

    # @param {int} curtTime an integer
    # @param {int} key an integer
    # @param {int} delta an integer
    # @return an integer
    def incr(self, curtTime, key, delta):
        # Write your code here
        if key not in self.client:
            return INT_MAX

        self.client[key].value += delta
        return self.get(curtTime, key)

    # @param {int} curtTime an integer
    # @param {int} key an integer
    # @param {int} delta an integer
    # @return an integer
    def decr(self, curtTime, key, delta):
        # Write your code here
        if key not in self.client:
            return INT_MAX
        self.client[key] -= delta
        return self.get(curtTime, key)
